File: Python/Binary Search Trees/17_least_greater_element.py
"""
Idea:
using BST: inserting value from vector backwards,as nodes. & keeping track of each node's successor
"""

from typing import List


# Define the Node class for the binary search tree
class Node:
    def __init__(self, val):
        self.data = val  # Initialize the data for the node
        self.left = None  # Initialize the left child as None
        self.right = None  # Initialize the right child as None


# Define the insert function to insert nodes into the binary search tree
def insert(root, val, succ):
    if root is None:
        root = Node(val)  # Create a new node if the root is None
        return root

    if root.data > val:
        succ[0] = root  # Set the current node as the successor
        root.left = insert(root.left, val, succ)  # Recursively insert into the left subtree
    elif root.data <= val:
        root.right = insert(root.right, val, succ)  # Recursively insert into the right subtree

    return root  # Return the root of the updated tree


# Define the Solution class to find the least greater element in the array
class Solution:
    def findLeastGreater(self, n, arr):
        root = None  # Initialize the root of the binary search tree as None
        ans = [-1] * n  # Initialize the answer array with -1 for all elements

        # Traverse the array in reverse order
        for i in range(n - 1, -1, -1):
            succ = [None]  # Initialize the successor as None
            root = insert(root, arr[i], succ)  # Insert the current element into the binary search tree
            if succ[0] is not None:
                ans[i] = succ[0].data  # If successor exists, update the answer array

        return ans  # Return the answer array containing the least greater elements


# {
# Driver Code Starts


class IntArray:
    def __init__(self) -> None:
        pass

    def Input(self, n):
        arr = [int(i) for i in input().strip().split()]  # array input
        return arr

    def Print(self, arr):
        for i in arr:
            print(i, end=" ")
        print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        n = int(input())

        arr = IntArray().Input(n)

        obj = Solution()
        res = obj.findLeastGreater(n, arr)

        IntArray().Print(res)

# } Driver Code Ends

File: C++/Binary Search Trees/17_least_greater_element.cpp
/*
Idea:
using BST: inserting value from vector backwards,as nodes. & keeping track of each node's successor
*/

//{ Driver Code Starts
/* Driver program to test above function */

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//Back-end complete function Template for C++
// Define the Node structure representing a node in a binary search tree
struct Node
{
    int data;  // Data stored in the node
    Node* left;  // Pointer to the left child
    Node* right;  // Pointer to the right child

    // Constructor to initialize the Node with a value
    Node(int val)
    {
        data = val;  // Set the data to the provided value
        left = NULL;  // Initialize the left child as NULL
        right = NULL;  // Initialize the right child as NULL
    }
};

// Function to insert a node into the binary search tree
Node* insert(Node*& root, int val, Node*& succ)
{
    // If the root is NULL, create a new node and return it
    if (root == NULL)
        return root = new Node(val);

    // If the value is less than the root's data, go to the left subtree
    if (root->data > val)
    {
        succ = root;  // Set the current node as the successor
        root->left = insert(root->left, val, succ);  // Recursively insert into the left subtree
    }
    // If the value is greater than or equal to the root's data, go to the right subtree
    else if (root->data <= val)
        root->right = insert(root->right, val, succ);  // Recursively insert into the right subtree

    return root;  // Return the root of the updated tree
}

class Solution {
public:
    // Function to find the least greater element for each element in the array
    vector<int> findLeastGreater(vector<int>& arr, int n) {
        Node* root = NULL;  // Initialize the root of the binary search tree as NULL
        vector<int> ans(n, -1);  // Initialize the answer vector with -1 for all elements

        // Traverse the array in reverse order
        for (int i = n - 1; i >= 0; i--)
        {
            Node* succ = NULL;  // Initialize the successor as NULL
            root = insert(root, arr[i], succ);  // Insert the current element into the binary search tree
            if (succ != NULL)
                ans[i] = succ->data;  // If successor exists, update the answer vector
        }

        return ans;  // Return the answer vector containing the least greater elements
    }
};


//{ Driver Code Starts.
int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n;
        cin >> n;
        vector<int>arr(n);
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        Solution ob;
        vector<int> ans = ob.findLeastGreater(arr, n);
        for (auto it : ans) {
            cout << it << " ";
        }
        cout << endl;
    }
    return 0;
}

// } Driver Code Ends